mlir-sycl/lib/Dialect/SYCL/Transforms/CPP.cpp

```diff
@@ -0,0 +1,212 @@
+/// functions receive as a parameter.
+class ConstantExplicitArg {
+public:
+  enum class Kind { ConstantArithArg };
+
+  unsigned getIndex() const { return index; }
+  Kind getKind() const { return kind; }
+
+protected:
+  ConstantExplicitArg(Kind kind, unsigned index) : kind(kind), index(index) {}
+
+private:
+  Kind kind;
+  unsigned index;
+};
+
+/// Class representing a constant arithmetic argument.
+class ConstantArithArg : public ConstantExplicitArg {
+public:
+  ConstantArithArg(unsigned index, Operation *definingOp)
+      : ConstantExplicitArg(ConstantExplicitArg::Kind::ConstantArithArg, index),
+        definingOp(definingOp) {
+    assert(definingOp && "Expecting valid operation");
+    assert(definingOp->hasTrait<OpTrait::ConstantLike>() &&
+           "Expecting constant");
+  }
+
+  /// Propagate the constant this argument represents.
+  void propagate(OpBuilder builder, Region &region) const;
+
+  static bool classof(const ConstantExplicitArg *c) {
+    return c->getKind() == ConstantExplicitArg::Kind::ConstantArithArg;
+  }
+
+private:
+  Operation *definingOp;
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// ConstantPropagationPass
+//===----------------------------------------------------------------------===//
+
+void ConstantArithArg::propagate(OpBuilder builder, Region &region) const {
+  Value toReplace = region.getArgument(getIndex());
+  toReplace.replaceAllUsesWith(builder.clone(*definingOp)->getResult(0));
+}
+
+auto ConstantPropagationPass::getConstantArgs(SYCLHostScheduleKernel op) {
+  auto isConstant = m_Constant();
+  // Map each argument to a ConstantExplicitArg and filter-out nullptr
+  // (non-const) ones.
+  return llvm::make_filter_range(
+      llvm::map_range(llvm::enumerate(op.getArgs()),
+                      [&](auto iter) -> std::unique_ptr<ConstantExplicitArg> {
+                        Value value = iter.value();
+                        if (matchPattern(value, isConstant)) {
+                          LLVM_DEBUG(llvm::dbgs()
+                                     << "Arith constant: " << value
+                                     << " at pos " << iter.index() << "\n");
+                          return std::make_unique<ConstantArithArg>(
+                              iter.index(), value.getDefiningOp());
+                        }
+                        return nullptr;
+                      }),
+      llvm::identity<std::unique_ptr<ConstantExplicitArg>>());
+}
+
+template <typename RangeTy>
+void ConstantPropagationPass::dropConstantArgs(RangeTy constants,
+                                               Builder &builder,
+                                               SYCLHostScheduleKernel launch) {
+  auto funcOp = static_cast<FunctionOpInterface>(getOperation());
+  llvm::BitVector launchIndices(launch->getNumOperands());
+  llvm::BitVector funcIndices(funcOp.getNumArguments());
```

Comment:
TODO(@jopperm): Flatten accessors etc. when raising the `schedule_kernel` op; otherwise the number of arguments won't match here.

Reply:
I should take that into account instead of flattening. I can address in the next patch in the series.